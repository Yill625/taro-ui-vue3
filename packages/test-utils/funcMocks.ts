import * as utils from "@taro-ui-vue3/utils/common"
import { pxTransformMockFn, getEnvMockFn } from './@tarojs/taro'

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let pxTransform: jest.SpyInstance

function initMocks() {
  beforeEach(() => {
    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    pxTransform.mockRestore()
  })
}

export default initMocks