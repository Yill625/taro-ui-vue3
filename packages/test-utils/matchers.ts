import * as utils from "@taro-ui-vue3/utils/common"

jest.mock('@taro-ui-vue3/utils/common')

let warn: jest.SpyInstance
let uuid: jest.SpyInstance

const asserted: Set<string> = new Set()
const NOOP = () => { }

// Ensure console.warn is called with expected message
const toHaveBeenWarned: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed = warn.mock.calls.some(args => args[0].indexOf(received) > -1)
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned` +
        (msgs.length
          ? `.\n\nActual messages:\n\n - ${msgs}`
          : ` but no warning was recorded.`)
    }
  }
}

// Ensure previous console.warn is called with expected message
const toHaveBeenWarnedLast: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed =
    warn.mock.calls[warn.mock.calls.length - 1][0].indexOf(received) > -1
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned last.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned last.\n\nActual messages:\n\n - ${msgs}`
    }
  }
}

// Ensure console.warn is called n times with expected message
const toHaveBeenWarnedTimes: jest.CustomMatcher = (received: string, n: number) => {
  asserted.add(received)
  let found = 0
  warn.mock.calls.forEach(args => {
    if (args[0].indexOf(received) > -1) {
      found++
    }
  })

  if (found === n) {
    return {
      pass: true,
      message: () => `expected "${received}" to have been warned ${n} times.`
    }
  } else {
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned ${n} times but got ${found}.`
    }
  }
}

function initWarnMatchers() {
  beforeEach(() => {
    asserted.clear()
    warn = jest.spyOn(console, 'warn').mockImplementation(NOOP)
    uuid = jest.spyOn(utils, 'uuid').mockReturnValue('__tests__2021')

    // NOT working in concurrent mode
    // should avoid use in test.concurrent()
    expect.extend({
      toHaveBeenWarned,
      toHaveBeenWarnedLast,
      toHaveBeenWarnedTimes
    })
  })

  afterEach(() => {
    const assertedArray = Array.from(asserted)
    const nonAssertedWarnings = warn.mock.calls
      .map(args => args[0])
      .filter(received => {
        return !assertedArray.some(assertedMsg => {
          return received.indexOf(assertedMsg) > -1
        })
      })

    uuid.mockRestore()
    warn.mockRestore()

    if (nonAssertedWarnings.length) {
      throw new Error(
        `test case threw unexpected warnings:\n - ${nonAssertedWarnings.join(
          '\n - '
        )}`
      )
    }
  })
}

export default initWarnMatchers