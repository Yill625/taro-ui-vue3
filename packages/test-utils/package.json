{
  "name": "@taro-ui-vue3/test-utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "license": "MIT",
  "dependencies": {
    "vue": "^3.2.0",
    "@vue/test-utils": "^2.0.0-rc.0"
  }
}