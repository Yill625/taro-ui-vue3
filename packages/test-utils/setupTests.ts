import toHaveBeenWarnedInit from './been-warned'

toHaveBeenWarnedInit()