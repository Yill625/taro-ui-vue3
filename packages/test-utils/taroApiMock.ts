import * as utils from "@taro-ui-vue3/utils/common"
import Taro from '@tarojs/taro'
import { createAnimation } from '@tarojs/taro-h5'

import {
  getEnvMockFn,
  pxTransformMockFn,
  chooseImageMockFn,
  querySelectorMockFn
} from './@tarojs/taro'

Taro.pxTransform = pxTransformMockFn
Taro.chooseImage = chooseImageMockFn
Taro.createSelectorQuery = querySelectorMockFn()
Taro.createAnimation = createAnimation

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let getenv: jest.SpyInstance
let pxTransform: jest.SpyInstance
let chooseImage: jest.SpyInstance

function initTaroMockApis() {
  beforeEach(() => {
    jest.mock('@tarojs/taro', () => {
      return Taro
    })

    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    getenv = jest
      .spyOn(Taro, 'getEnv')
      .mockImplementation(getEnvMockFn)

    chooseImage = jest
      .spyOn(Taro, 'chooseImage')
      .mockImplementation(chooseImageMockFn)

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    getenv.mockRestore()
    pxTransform.mockRestore()
    chooseImage.mockRestore()
  })
}

export default initTaroMockApis
