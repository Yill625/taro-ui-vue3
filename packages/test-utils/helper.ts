import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}