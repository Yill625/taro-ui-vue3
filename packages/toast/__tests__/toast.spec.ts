import { genMountFn } from '@taro-ui-vue3/test-utils/helper'
import AtToast from '../index.vue'
import statusImg from '../img.json'

const mountFn = genMountFn(AtToast)

const icon = 'loading'
const text = '测试数据'
const image = 'http://img.com/06540.png'

describe('AtToast', () => {
  it('should render toast using image and match snapshot', () => {
    const wrapper = mountFn({
      isOpened: true,
      hasMask: true,
      text,
      image
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render toast using icon and match snapshot', () => {
    const wrapper = mountFn({
      isOpened: true,
      hasMask: true,
      text,
      icon
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render nothing if toast is not opened', () => {
    const wrapper = mountFn()
    expect(wrapper.find('.at-toast').exists()).toBeFalsy()
  })

  it('should render prop text', () => {
    const wrapper = mountFn({ isOpened: true, text })
    const textEl = wrapper.find('.toast-body-content__info')
    expect(textEl.exists()).toBeTruthy()
    expect(textEl.text()).toEqual(text)
  })

  it('should render prop hasMask', () => {
    const wrapper = mountFn({ isOpened: true, hasMask: true })
    expect(wrapper.find('.at-toast__overlay').exists()).toBeTruthy()
  })

  it.each([
    'loading',
    'error',
    'success'
  ])('should render prop status -- %s and use builtin status image', async (status) => {
    const wrapper = mountFn({ isOpened: true, status })
    expect(
      wrapper.find('.toast-body').classes()
    ).toContain(`at-toast__body--${status}`)

    const imageEl = wrapper.find('.toast-body-content__img-item')
    expect(imageEl.exists()).toBeTruthy()
    expect(imageEl.attributes('src')).toEqual(statusImg[status])
  })
})

describe('AtToast events ', () => {
  beforeEach(() => {
    jest.useFakeTimers()
    jest.runOnlyPendingTimers()
  })
  it('should emit close event and close Toast', async () => {
    const onClose = jest.fn()
    const wrapper = mountFn({ isOpened: true, onClose: onClose })
    expect(wrapper.find(".at-toast").exists()).toBeTruthy()

    await wrapper.find('.at-toast .toast-body').trigger('tap')
    await wrapper.vm.$nextTick()

    expect(wrapper.emitted()).toHaveProperty('close')
    expect(onClose).toBeCalled()
    expect(wrapper.find(".at-toast").exists()).toBeFalsy()
  })

  it.each([
    3000,
    1000,
    0
  ])('should be closed when duration time over -- %d ms', async (duration) => {
    const wrapper = mountFn({ isOpened: true, duration })

    expect(wrapper.vm.duration).toEqual(duration)
    expect(wrapper.find(".at-toast").exists()).toBeTruthy()

    jest.advanceTimersByTime(duration)
    await wrapper.vm.$nextTick()

    expect(wrapper.find(".at-toast").exists()).toBeFalsy()
  })

  it('should emit click and close events', async () => {
    const onClick = jest.fn()
    const onClose = jest.fn()
    const wrapper = mountFn({ isOpened: true, onClick, onClose })
    expect(wrapper.find(".at-toast").exists()).toBeTruthy()

    await wrapper.find('.at-toast .toast-body').trigger('tap')

    expect(wrapper.emitted()).toHaveProperty('click')
    expect(wrapper.emitted()).toHaveProperty('close')
    expect(onClick).toBeCalled()
    expect(onClose).toBeCalled()

    await wrapper.vm.$nextTick()
    expect(wrapper.find(".at-toast").exists()).toBeFalsy()
  })

  it('should not mit click and close events if status is loading', async () => {
    const onClick = jest.fn()
    const onClose = jest.fn()
    const wrapper = mountFn({ isOpened: true, status: 'loading', onClick, onClose })
    expect(wrapper.find(".at-toast").exists()).toBeTruthy()

    await wrapper.find('.at-toast .toast-body').trigger('tap')

    expect(wrapper.emitted()).not.toHaveProperty('click')
    expect(wrapper.emitted()).not.toHaveProperty('close')
    expect(onClick).not.toBeCalled()
    expect(onClose).not.toBeCalled()

    await wrapper.vm.$nextTick()
    expect(wrapper.find(".at-toast").exists()).toBeTruthy()
  })
})
