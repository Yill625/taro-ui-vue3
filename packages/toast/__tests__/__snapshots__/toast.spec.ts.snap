// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtToast should render toast using icon and match snapshot 1`] = `
<view
  class="at-toast"
>
  <view
    class="at-toast__overlay"
  />
  <view
    class="toast-body"
  >
    <view
      class="toast-body-content"
    >
      <view
        class="toast-body-content__icon"
      >
        <text
          class="at-icon at-icon-loading"
        />
      </view>
      <view
        class="toast-body-content__info"
      >
        <text>
          测试数据
        </text>
      </view>
    </view>
  </view>
</view>
`;

exports[`AtToast should render toast using image and match snapshot 1`] = `
<view
  class="at-toast"
>
  <view
    class="at-toast__overlay"
  />
  <view
    class="toast-body at-toast__body--custom-image"
  >
    <view
      class="toast-body-content"
    >
      <view
        class="toast-body-content__img"
      >
        <image
          class="toast-body-content__img-item"
          mode="scaleToFill"
          src="http://img.com/06540.png"
        />
      </view>
      <view
        class="toast-body-content__info"
      >
        <text>
          测试数据
        </text>
      </view>
    </view>
  </view>
</view>
`;
